File: app/src/main/java/feelings/guide/profile/LocaleUtil.java
package feelings.guide.profile;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.preference.PreferenceManager;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;

public class LocaleUtil {
    private static final List<String> SUPPORTED_LANGUAGES = Arrays.asList(
            Locale.ENGLISH.getLanguage(),
            new Locale("ru").getLanguage());

    static final String SELECTED_LANGUAGE = "Feelings.Guide.Selected.Language";

    public static Context setLocale(Context context) {
        String defaultLanguage = Locale.getDefault().getLanguage();
        if (!SUPPORTED_LANGUAGES.contains(defaultLanguage)) {
            defaultLanguage = Locale.ENGLISH.getLanguage();
        }

        String lang = getPersistedData(context, defaultLanguage);
        return setLocale(context, lang);
    }

    private static Context setLocale(Context context, String language) {
        persist(context, language);

        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
            ? updateResources(context, language)
            : updateResourcesLegacy(context, language);
    }

    private static String getPersistedData(Context context, String defaultLanguage) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getString(SELECTED_LANGUAGE, defaultLanguage);
    }

    private static void persist(Context context, String language) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor editor = preferences.edit();

        editor.putString(SELECTED_LANGUAGE, language);
        editor.apply();
    }

    @TargetApi(Build.VERSION_CODES.N)
    private static Context updateResources(Context context, String language) {
        Locale locale = new Locale(language);
        Locale.setDefault(locale);

        Configuration configuration = context.getResources().getConfiguration();
        configuration.setLocale(locale);
        configuration.setLayoutDirection(locale);

        return context.createConfigurationContext(configuration);
    }

    @SuppressWarnings("deprecation")
    private static Context updateResourcesLegacy(Context context, String language) {
        Locale locale = new Locale(language);
        Locale.setDefault(locale);

        Resources resources = context.getResources();

        Configuration configuration = resources.getConfiguration();
        configuration.locale = locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            configuration.setLayoutDirection(locale);
        }

        resources.updateConfiguration(configuration, resources.getDisplayMetrics());

        return context;
    }
}

File: app/src/main/kotlin/feelings/guide/profile/LocaleUtil.kt
package feelings.guide.profile

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import androidx.preference.PreferenceManager
import feelings.guide.ui.settings.SELECTED_LANGUAGE_KEY
import java.util.*

object LocaleUtil {
    private val SUPPORTED_LANGUAGES = listOf(Locale.ENGLISH.language, Locale("ru").language)

    fun setLocale(context: Context): Context {
        var defaultLanguage = Locale.getDefault().language
        if (!SUPPORTED_LANGUAGES.contains(defaultLanguage)) {
            defaultLanguage = Locale.ENGLISH.language
        }

        val lang = getPersistedData(context, defaultLanguage)
        return if (lang != null)
            setLocale(context, lang)
        else context
    }

    private fun setLocale(context: Context, language: String): Context {
        persist(context, language)

        return if (Build.VERSION.SDK_INT > Build.VERSION_CODES.N)
            updateResources(context, language)
        else
            updateResourcesLegacy(context, language)
    }

    private fun getPersistedData(context: Context, defaultLanguage: String): String? {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getString(SELECTED_LANGUAGE_KEY, defaultLanguage)
    }

    private fun persist(context: Context, language: String?) {
        PreferenceManager.getDefaultSharedPreferences(context)
            .edit()
            .putString(SELECTED_LANGUAGE_KEY, language)
            .apply()
    }

    @TargetApi(Build.VERSION_CODES.N)
    private fun updateResources(context: Context, language: String): Context {
        val locale = Locale(language)
        Locale.setDefault(locale)

        val configuration = context.resources.configuration
        configuration.setLocale(locale)
        configuration.setLayoutDirection(locale)

        return context.createConfigurationContext(configuration)
    }

    @Suppress("DEPRECATION")
    private fun updateResourcesLegacy(context: Context, language: String): Context {
        val locale = Locale(language)
        Locale.setDefault(locale)

        val resources = context.resources

        val configuration = resources.configuration
        configuration.locale = locale
        configuration.setLayoutDirection(locale)

        resources.updateConfiguration(configuration, resources.displayMetrics)

        return context
    }
}
